cmd/rpc/sock.go

```diff
@@ -0,0 +1,455 @@
+	}
+	// if rootHeight is the same as the RootChainInfo height
+	if rootHeight == sub.Info.Height || rootHeight == 0 {
+		// exit with a copy the validator set
+		return lib.NewValidatorSet(sub.Info.ValidatorSet)
+	}
+	// if rootHeight is 1 before the RootChainInfo height
+	if rootHeight == sub.Info.Height-1 {
+		// exit with a copy of the previous validator set
+		return lib.NewValidatorSet(sub.Info.LastValidatorSet)
+	}
+	// warn of the remote RPC call to the root chain API
+	r.log.Warnf("Executing remote GetValidatorSet call with requested height=%d for rootChainId=%d", rootHeight, rootChainId)
+	// execute the remote RPC call to the root chain API
+	return sub.ValidatorSet(rootHeight, id)
+}
+
+// GetOrders() returns the order book from the root-chain
+func (r *RCManager) GetOrders(rootChainId, rootHeight, id uint64) (*lib.OrderBook, lib.ErrorI) {
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return nil, lib.ErrNotSubscribed()
+	}
+	// if the root chain id and height is the same as the info
+	if sub.Info.Height == rootHeight {
+		// exit with the order books from memory
+		return sub.Info.Orders, nil
+	}
+	// warn of the remote RPC call to the root chain API
+	r.log.Warnf("Executing remote GetOrders call with requested height=%d for rootChainId=%d", rootHeight, rootChainId)
+	// execute the remote call
+	books, err := sub.Orders(rootHeight, id)
+	// if an error occurred during the remote call
+	if err != nil {
+		// exit with error
+		return nil, err
+	}
+	// ensure the order book isn't empty
+	if books == nil || len(books.OrderBooks) == 0 {
+		// exit with error
+		return nil, lib.ErrEmptyOrderBook()
+	}
+	// exit with the first (and only) order book in the list
+	return books.OrderBooks[0], nil
+}
+
+// Order() returns a specific order from the root order book
+func (r *RCManager) GetOrder(rootChainId, height, orderId, chainId uint64) (*lib.SellOrder, lib.ErrorI) {
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return nil, lib.ErrNotSubscribed()
+	}
+	return sub.Order(height, orderId, chainId)
+}
+
+// IsValidDoubleSigner() returns if an address is a valid double signer for a specific 'double sign height'
+func (r *RCManager) IsValidDoubleSigner(rootChainId, height uint64, address string) (*bool, lib.ErrorI) {
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return nil, lib.ErrNotSubscribed()
+	}
+	// warn of the remote RPC call to the API of the 'root chain'
+	r.log.Warnf("Executing remote IsValidDoubleSigner call for address=%s at double sign height=%d and rootChainId=%d", address, height, rootChainId)
+	// exit with the results of the remote RPC call to the API of the 'root chain'
+	return sub.IsValidDoubleSigner(height, address)
```

Comment:
Here and in `GetCheckpoint` from the looks of the call there's no other way to retrieve the information besides doing a http request so a warning is given, for most of the other functions who do ultimately fall back to http when the data is not present it makes sense but this one seems like a false positive unless is expected for this to be converted to ws information too and if so it should be documented